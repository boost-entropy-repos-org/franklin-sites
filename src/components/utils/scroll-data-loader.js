const scrollOffsetPercentage = 10;
const scrollOffsetFactor = 1 + scrollOffsetPercentage / 100;

export const isNotScrollable = (node) => {
  const { scrollHeight, offsetHeight } = node.current;
  return scrollHeight <= offsetHeight * scrollOffsetFactor;
};

export const isBottom = (node) => {
  const { scrollHeight, scrollTop, offsetHeight } = node.current;
  return scrollHeight - Math.ceil(scrollTop) <= offsetHeight * scrollOffsetFactor;
};

export const checkLoadMoreItems = (loading, hasMoreData, ref, setLoadMoreItems) => {
  if (!loading && hasMoreData && isBottom(ref)) {
    setLoadMoreItems(true);
  }
};

export const checkOnDataLoad = (
  hasMoreData,
  setLoading,
  onLoadMoreItems,
  setLoadMoreItems,
  ref,
) => {
  if (hasMoreData && isNotScrollable(ref)) {
    setLoading(true);
    onLoadMoreItems();
  } else {
    setLoading(false);
    setLoadMoreItems(false);
  }
};

export const checkOnLoadMoreItems = (loadMoreItems, setLoading, onLoadMoreItems) => {
  if (loadMoreItems) {
    setLoading(true);
    onLoadMoreItems();
  }
};
