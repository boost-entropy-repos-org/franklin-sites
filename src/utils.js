export function getLastIndexOfSubstringIgnoreCase(string, substring) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export const getFlattenedPaths = (currentItems, value, path = []) => {
  let flattened = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, value, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!value || thisNode.value === value) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

function getSquashedPath(items, sep = ' / ') {
  return {
    value: items[items.length - 1].value,
    pathLabel: items.map(item => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
    items,

  };
}

export function getSquashedArrayOfPaths(paths) {
  // console.log(paths);
  return paths.map(items => getSquashedPath(items));
}
