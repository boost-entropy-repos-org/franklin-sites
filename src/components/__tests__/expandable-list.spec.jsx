import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ExpandableList, { ExpandableMessage } from '../expandable-list';
import { fillArray } from '../../utils';

const descriptionString = 'foo';

describe('ExpandableMessage', () => {
  let setExpanded;
  beforeEach(() => {
    setExpanded = jest.fn();
  });

  test('should render', () => {
    const { asFragment } = render(
      <ExpandableMessage
        expanded={false}
        setExpanded={setExpanded}
        descriptionString={descriptionString}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should call setExpanded when button is clicked and button text be "More descriptionString"', () => {
    const { container } = render(
      <ExpandableMessage
        expanded={false}
        setExpanded={setExpanded}
        descriptionString={descriptionString}
      />,
    );
    const button = container.querySelector('button');
    fireEvent.click(button);
    expect(setExpanded).toHaveBeenCalled();
    expect(button.textContent).toEqual(`More ${descriptionString}`);
  });

  test('button text should be "Less descriptionString"', () => {
    const { container } = render(
      <ExpandableMessage
        expanded
        setExpanded={setExpanded}
        descriptionString={descriptionString}
      />,
    );
    const button = container.querySelector('button');
    fireEvent.click(button);
    expect(setExpanded).toHaveBeenCalled();
    expect(button.textContent).toEqual(`Less ${descriptionString}`);
  });
});

describe('ExpandableList', () => {
  const numberItems = 10;
  const numberCollapsedItems = 5;
  const items = fillArray(numberItems, (element, index) => ({
    id: `id${index}`,
    content: `some content ${index}`,
  }));

  test('should render', () => {
    const { asFragment } = render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {items}
      </ExpandableList>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should initially have numberCollapsedItems and then total number of items after More button click ', () => {
    const { container } = render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {items}
      </ExpandableList>,
    );
    let listItems = container.querySelectorAll('li');
    expect(listItems).toHaveLength(numberCollapsedItems);
    const button = container.querySelector('button');
    fireEvent.click(button);
    listItems = container.querySelectorAll('li');
    expect(listItems).toHaveLength(numberItems);
  });
});
