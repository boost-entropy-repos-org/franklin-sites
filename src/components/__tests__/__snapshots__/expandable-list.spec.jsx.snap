// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableList should render 1`] = `
<DocumentFragment>
  <ul
    class="no-bullet"
  >
    <li>
      some content 0
    </li>
    <li>
      some content 1
    </li>
    <li>
      some content 2
    </li>
    <li>
      some content 3
    </li>
    <li>
      some content 4
    </li>
    <button
      class="button link-button"
      data-testid="expandable-message"
      type="button"
    >
      More foo
    </button>
  </ul>
</DocumentFragment>
`;

exports[`ExpandableMessage should render 1`] = `
<DocumentFragment>
  <button
    class="button link-button"
    data-testid="expandable-message"
    type="button"
  >
    More foo
  </button>
</DocumentFragment>
`;
